src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -152,86 +84,55 @@ private static int GetEngineCount()
         private int _eventQueueProcessingRequested;
 
         //
-        // True if we've reached the handle value limit for this event port, and thus must allocate a new event port
-        // on the next handle allocation.
+        // Registers the Socket with a SocketAsyncEngine, and returns the associated engine.
         //
-        private bool IsFull { get { return _nextHandle == MaxHandles; } }
+        public static SocketAsyncEngine RegisterSocket(IntPtr socketHandle, SocketAsyncContext context)
+        {
+            int engineIndex = Math.Abs(Interlocked.Increment(ref s_allocateFromEngine) % s_engines.Length);
+            SocketAsyncEngine engine = s_engines[engineIndex];
+            engine.RegisterCore(socketHandle, context);
+            return engine;
+        }
 
-        //
-        // Allocates a new {SocketAsyncEngine, handle} pair.
-        //
-        private static void AllocateToken(SocketAsyncContext context, out SocketAsyncEngine? engine, out IntPtr handle)
+        private void RegisterCore(IntPtr socketHandle, SocketAsyncContext context)
         {
-            lock (s_lock)
+            try
             {
-                engine = s_currentEngines[s_allocateFromEngine];
-                if (engine == null)
+                bool added = _handleToContextMap.TryAdd(socketHandle, new SocketAsyncContextWrapper(context));
+                Debug.Assert(added, "Add should always succeed");
+
+                Interop.Error error = Interop.Sys.TryChangeSocketEventRegistration(_port, socketHandle, Interop.Sys.SocketEvents.None,
+                    Interop.Sys.SocketEvents.Read | Interop.Sys.SocketEvents.Write, socketHandle);
```

Comment:
Try allocating a GCHandle for the context `GCHandle handle = GCHandle.Alloc(context);` and use that handle instead of `socketHandle`.

Then you can drop `ConcurrentDictionary<IntPtr, SocketAsyncContextWrapper>` and use `.Target` instead to get the context back e.g. `context = Unsafe.As<SocketAsyncContext>(((GCHandle)handle).Target))`

Assuming `GCHandle.Target` is faster than the `ConcurrentDictionary` TryGet anyway? /cc @stephentoub 

Would want to also swap `_handleToContextMap.TryRemove` for `handle.Free()`

Reply:
> Assuming GCHandle.Target is faster than the ConcurrentDictionary TryGet anyway?

Would need to measure.

> Unsafe.As<SocketAsyncContext>(((GCHandle)handle).Target))

Not with an unsafe cast! Too many things could go wrong for me to feel comfortable with that, e.g. a race between the socket getting finalized and an event for it getting processed.

Reply:
I'm not going to pursue this as part of this PR.